File: backend/app/api/tasks.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.api.dependencies import get_current_user
from app.database.dependencies import get_db
from app.models.user import User
from app.schemas.task import (
    TaskCreate,
    TaskResponse,
    TaskUpdate,
)
from app.services.task_service import (
    create_task,
    delete_task,
    get_task,
    get_tasks,
    update_task,
)

router = APIRouter(
    prefix="/tasks",
    tags=["Tasks"],
)

@router.post(
    "",
    response_model=TaskResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_new_task(
    data: TaskCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return create_task(
        db=db,
        user_id=current_user.id,
        data=data,
    )

@router.get(
    "",
    response_model=list[TaskResponse],
)
def list_tasks(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return get_tasks(
        db=db,
        user_id=current_user.id,
    )

@router.get(
    "/{task_id}",
    response_model=TaskResponse,
)
def get_single_task(
    task_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    task = get_task(
        db=db,
        user_id=current_user.id,
        task_id=task_id,
    )

    if task is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found.",
        )

    return task

@router.patch(
    "/{task_id}",
    response_model=TaskResponse,
)
def update_existing_task(
    task_id: int,
    data: TaskUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    task = get_task(
        db=db,
        user_id=current_user.id,
        task_id=task_id,
    )

    if task is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found.",
        )

    return update_task(
        db=db,
        task=task,
        data=data,
    )

@router.delete(
    "/{task_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_existing_task(
    task_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    task = get_task(
        db=db,
        user_id=current_user.id,
        task_id=task_id,
    )

    if task is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found.",
        )

    delete_task(
        db=db,
        task=task,
    )

    return None

File: frontend/src/api/tasks.ts
import apiClient from "./client";

export interface Task {
  id: number;
  user_id: number;
  title: string;
  description?: string | null;
  priority: string;
  due_at?: string | null;
  status: string;
  completed_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface CreateTaskData {
  title: string;
  description?: string;
  priority?: string;
  due_at?: string;
}

export interface UpdateTaskData {
  title?: string;
  description?: string;
  priority?: string;
  due_at?: string;
  status?: string;
}

export async function createTask(
  data: CreateTaskData
): Promise<Task> {
  const response = await apiClient.post<Task>(
    "/tasks",
    data
  );

  return response.data;
}

export async function getTasks(): Promise<Task[]> {
  const response = await apiClient.get<Task[]>("/tasks");

  return response.data;
}

export async function getTask(
  taskId: number
): Promise<Task> {
  const response = await apiClient.get<Task>(
    `/tasks/${taskId}`
  );

  return response.data;
}

export async function updateTask(
  taskId: number,
  data: UpdateTaskData
): Promise<Task> {
  const response = await apiClient.patch<Task>(
    `/tasks/${taskId}`,
    data
  );

  return response.data;
}

export async function deleteTask(
  taskId: number
): Promise<void> {
  await apiClient.delete(`/tasks/${taskId}`);
}
